internal/app/backend/backend_service.go

```diff
@@ -330,6 +339,64 @@ func (s *backendService) AssignTickets(ctx context.Context, req *pb.AssignTicket
 	return resp, nil
 }
 
+func handleMatch(ctx context.Context, match *pb.Match, store statestore.Service) error {
+	backfill := match.GetBackfill()
+
+	if backfill == nil {
+		return nil
+	}
+
+	ticketdIds := make([]string, len(match.Tickets))
+
+	for _, t := range match.Tickets {
+		ticketdIds = append(ticketdIds, t.Id)
+	}
+
+	if backfill.Id != "" {
+		return doUpdateBackfill(ctx, backfill, ticketdIds, store)
+	}
+
+	return doCreateBackfill(ctx, backfill, ticketdIds, store)
+}
+
+func doCreateBackfill(ctx context.Context, backfill *pb.Backfill, ticketIds []string, store statestore.Service) error {
+	backfill.Id = xid.New().String()
+	backfill.CreateTime = ptypes.TimestampNow()
+	return store.CreateBackfill(ctx, backfill, ticketIds)
+}
+
+func doUpdateBackfill(ctx context.Context, backfill *pb.Backfill, ticketIds []string, store statestore.Service) error {
+	m := store.NewMutex(backfill.Id)
+	err := m.Lock(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		_, unlockErr := m.Unlock(ctx)
+
+		if unlockErr != nil {
+			logger.WithFields(logrus.Fields{"backfill_id": backfill.Id}).WithError(unlockErr).Error("failed to make unlock")
+		}
+	}()
+
+	bf, ids, err := store.GetBackfill(ctx, backfill.Id)
+
+	if err != nil {
+		return err
+	}
+
+	if bf.Generation != backfill.Generation {
+		return errors.New("backfill generation mismatch")
```

Comment:
I think if you specify exact generations here in a log message it will be a bit more informative. 
Like: fmt.Errorf("backfill generation mismatch, current %s, incoming %s", ...)